tiledb/sm/array/consistency.h

```diff
@@ -0,0 +1,192 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines classes ConsistencyController and ConsistencySentry.
+ */
+
+#ifndef TILEDB_CONSISTENCY_H
+#define TILEDB_CONSISTENCY_H
+
+#include <iterator>
+#include <map>
+
+#include "tiledb/common/common.h"
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/filesystem/uri.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm {
+
+class Array;
+class ConsistencySentry;
+
+/**
+ * Tracks the open arrays on disk, considering that a given
+ * URI can have multiple open arrays.
+ *
+ * Intended to act as a singleton global.
+ * There is only one (global) copy of this class used in practice.
+ */
+class ConsistencyController {
+  /**
+   * Makes available the private APIs of ConsistencyController for array
+   * registration and deregistration.
+   */
+  friend class ConsistencySentry;
+  /**
+   * Makes available internals of ConsistencyController for testing.
+   */
+  friend class WhiteboxConsistencyController;
+
+ public:
+  using entry_type = std::multimap<const URI, Array&>::const_iterator;
+
+  /**
+   * Constructor.
+   *
+   * Default constructed as the initial state of the singleton Controller with
+   * no registered arrays.
+   * Note: nothing can be registered in the singleton until an array is opened.
+   */
+  ConsistencyController() = default;
+
+  /** Copy Constructor is deleted. */
+  ConsistencyController(const ConsistencyController&) = delete;
+
+  /** Move Constructor is deleted. */
+  ConsistencyController(ConsistencyController&&) = delete;
+
+  /** Copy assignment is deleted. */
+  ConsistencyController& operator=(const ConsistencyController&) = delete;
+
+  /** Move assignment is deleted. */
+  ConsistencyController& operator=(ConsistencyController&&) = delete;
+
+  /** Destructor. */
+  ~ConsistencyController() = default;
+
+  /**
+   * Register the given Array as open.
+   *
+   * @return Sentry object whose lifespan is the same as the registration.
+   */
+  ConsistencySentry make_sentry(const URI uri, Array& array);
+
+  /** Returns true if the array is open, i.e. registered in the multimap. */
+  bool is_open(const URI uri);
+
+  /**
+   * Returns true if the given URIs have the same "prefix" and could
+   * potentially intersect one another.
+   * i.e. The second URI is an element of the first (or vice versa).
+   *
+   * Note: The order of the arguments does not matter;
+   * the API is checking for working tree intersection.
+   */
+  bool is_element_of(const URI uri, const URI intersecting_uri);
```

Comment:
This seems generally useful, consider moving to `URI` in the future.

Reply:
We have the beginnings of where this goes in `tiledb/storage_format/uri`. As of today it only has a few functions extracted from `misc/util.h`.

Reply:
> should this be moved to `parse_uri.cc`?

In a later PR. This one already has suffered too much scope creep.